ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c

```diff
@@ -27,53 +27,68 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
-  iota_stor_pack_t pack;
+  iota_stor_pack_t approvers_pack;
+  iota_stor_pack_t approvees_pack;
   hash243_stack_t stack = NULL;
+  hash243_queue_t approvees = NULL;
   flex_trit_t *curr_tx_hash = NULL;
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, transaction_pack);
 
   *subtangle_size = 0;
 
-  if ((ret = hash_pack_init(&pack, 10)) != RC_OK) {
-    goto done;
-  }
-
-  if ((ret = hash243_stack_push(&stack, entry_point)) != RC_OK) {
-    goto done;
-  }
+  ERR_BIND_GOTO(hash_pack_init(&approvers_pack, 10), ret, done);
+  ERR_BIND_GOTO(hash_pack_init(&approvees_pack, 2), ret, done);
+  ERR_BIND_GOTO(hash243_stack_push(&stack, entry_point), ret, done);
 
   while (!hash243_stack_empty(stack)) {
     curr_tx_hash = hash243_stack_peek(stack);
 
     if (!hash_to_indexed_hash_set_map_contains(tx_to_approvers, curr_tx_hash)) {
-      hash_pack_reset(&pack);
-      if ((ret = iota_tangle_transaction_load_hashes_of_approvers(tangle, curr_tx_hash, &pack,
+      hash_pack_reset(&approvers_pack);
+      if ((ret = iota_tangle_transaction_load_hashes_of_approvers(tangle, curr_tx_hash, &approvers_pack,
                                                                   subtangle_before_timestamp)) != RC_OK) {
         log_error(logger_id, "Failed in loading approvers, error code is: %" PRIu64 "\n", ret);
         goto done;
       }
-      if ((ret = hash_to_indexed_hash_set_map_add_new_set(tx_to_approvers, curr_tx_hash, &curr_tx,
-                                                          (*subtangle_size)++)) != RC_OK) {
-        goto done;
-      }
+
+      ERR_BIND_GOTO(
+          hash_to_indexed_hash_set_map_add_new_set(tx_to_approvers, curr_tx_hash, &curr_tx, (*subtangle_size)++), ret,
+          done);
       hash243_stack_pop(&stack);
-      while (pack.num_loaded > 0) {
-        curr_tx_hash = ((flex_trit_t *)pack.models[--pack.num_loaded]);
-        // Add each found approver to the currently traversed tx
-        if ((ret = hash243_stack_push(&stack, curr_tx_hash)) != RC_OK) {
-          goto done;
+      while (approvers_pack.num_loaded > 0) {
+        curr_tx_hash = ((flex_trit_t *)approvers_pack.models[--approvers_pack.num_loaded]);
+
+        ERR_BIND_GOTO(iota_tangle_transaction_load_partial(tangle, curr_tx_hash, &transaction_pack,
+                                                           PARTIAL_TX_MODEL_ESSENCE_ATTACHMENT_METADATA),
+                      ret, done)
+        hash_pack_reset(&approvees_pack);
+
+        if (transaction_trunk(&tx) == transaction_branch(&tx)) {
+          hash243_queue_push(&approvees, transaction_branch(&tx));
+        } else {
+          hash243_queue_push(&approvees, transaction_branch(&tx));
+          hash243_queue_push(&approvees, transaction_trunk(&tx));
         }
-        if ((ret = hash243_set_add(&curr_tx->approvers, curr_tx_hash)) != RC_OK) {
-          goto done;
+
+        ERR_BIND_GOTO(iota_tangle_transaction_find(tangle, NULL, NULL, NULL, approvees, &approvees_pack), ret, done);
```

Comment:
This one was created for the API call `findTransactions` and is *really* slow, I'm not sure if we should use it in CW.